### ISSUE TITLE
A device assert the author explicitly kept is computed and then discarded

### ISSUE THREAD

**Issue by benkelaya** [bledden/triton-msl#8]
**Author:** Hocine Benkelaya (with NeuroBrix)

## The point that decides it

`tl.device_assert` reaches the IR **only** when `options.debug` is true:

```python
# triton/language/semantic.py
def device_assert(self, cond, msg, mask):
    if not self.builder.options.debug:
        return
    ...
    return self.tensor(self.builder.create_assert(cond.handle, msg), tl.void)
```

So a `tt.assert` arriving at a backend lowerer is, by construction, one the
author asked to keep — with `@triton.jit(debug=True)` or `TRITON_DEBUG`. The
release path is the frontend gate this op has already passed.

`generic_lowerer.py` elides it anyway, and the comment gives the reason:

> matching CUDA's RELEASE behavior (Triton compiles device asserts out unless
> TRITON_DEBUG)

That is the gate the op already went through. What is being overridden here is
not a release default; it is an explicit request.

## What is actually emitted

Eight floats. A kernel loads row `idx` of a two-row table, guarded by
`tl.device_assert(idx < ROWS)`:

| | `tt.assert` in the TTIR | generated MSL |
|---|---:|---:|
| `@triton.jit(debug=True)` | **2** | 836 bytes, **no trap of any kind** |
| `@triton.jit(debug=False)` | 0 | 481 bytes |

The `debug=True` body is not shorter for having dropped the guard. It is
**longer**:

```c
        bool mask_5 = (int)idx < (int)c_4;      // the author's assert predicate
        long r_6 = static_cast<long>(idx);
        long r_7 = r_6 * c_0;
        bool mask_8 = (long)r_7 <= (long)c_2;   // the int-overflow sanitizer's
        bool mask_9 = (long)r_7 >= (long)c_1;
        bool r_10 = mask_8 & mask_9;
        int r_11 = idx * c_3;                   // ...and none of the three is read
        float val_12 = static_cast<float>(table[(r_11 + lid)]);
        out[(0 + lid)] = val_12;
```

Five extra values, a 64-bit widening among them, computed and thrown away. The
guard is paid for in registers and instructions and is not delivered.

## Why this is not only a missing debug trap

The elision comment states:

> This is NOT a silent-wrong: it never alters a computed value, it only forgoes
> a debug trap on already-out-of-contract inputs.
> (A store/atomic — an output-AFFECTING side effect — is refused below.)

The premise holds only when the program is correct, which is the one case an
assert does not exist for. In the bounds-check idiom the assert is not a
diagnostic beside the access — it is the guard that makes the following access
legal. Dropping it does not leave the result unchanged; it lets the guarded
access run.

Measured on three ports of the same FlagGems-lineage kernels, whose only bounds
check was that assert:

| kernel | store mask | with the assert elided |
|---|---|---|
| index_select | `out_mask & index_valid_mask` | nothing written; the output element keeps the allocator's residue |
| embedding | `cols < N` only | **reads past the end of the weight** (4 floats past a 24-float tensor) |
| index_put | `off < N` only | **writes past the end of the tensor** (8 floats past) |

The store the comment says is "refused below" is not inside the assert — it is
*guarded by* it, and it is reached.

## Reproducer

```python
import torch, triton, triton.language as tl

N = 4

def build(debug):
    @triton.jit(debug=debug)
    def k(table, out, idx, N: tl.constexpr, ROWS: tl.constexpr):
        tl.device_assert(idx < ROWS, "row out of range")
        off = tl.arange(0, N)
        tl.store(out + off, tl.load(table + idx * N + off))
    return k

table = torch.arange(8, dtype=torch.float32, device="mps").reshape(2, N)
for debug in (True, False):
    out = torch.zeros(N, dtype=torch.float32, device="mps")
    h = build(debug)[(1,)](table, out, 0, N=N, ROWS=2)
    print(f"debug={debug}: tt.assert in TTIR = {h.asm['ttir'].count('tt.assert')}, "
          f"'assert' in MSL = {'assert' in h.asm['msl'].lower()}")

# row 7 of a two-row table: no exception, from either the launch or the sync
out = torch.full((N,), -99.0, dtype=torch.float32, device="mps")
build(True)[(1,)](table, out, 7, N=N, ROWS=2)
torch.mps.synchronize()
print("row 7 of a 2-row table ->", out.tolist())
```

Output on the machine below:

```
debug=True: tt.assert in TTIR = 2, 'assert' in MSL = False
debug=False: tt.assert in TTIR = 0, 'assert' in MSL = False
row 7 of a 2-row table -> [0.0, 0.0, 0.0, 0.0]
```

## What a fix could look like

Two shapes, and the choice between them is the maintainers' rather than mine:

1. **Honour it.** Apple GPUs have no device-side abort, but they do not need
   one for this: the predicate is *already computed*, so lowering `tt.assert`
   to a conditional store into a one-word status buffer that the launcher reads
   after completion gives the same host-visible semantics CUDA has. A
   conditional store survives this lowerer intact today — `if (mask) { buf[0] =
   code; }` is emitted verbatim — so the device half needs no new machinery;
   only the launcher would have to bind the buffer and raise on it, and only
   for kernels that carry an assert.

2. **Refuse, or at minimum announce.** If the semantics will not be honoured,
   silently dropping a guard the author explicitly asked for is the part worth
   changing. A compile-time diagnostic naming the kernel and the assert message
   would let a `debug=True` user learn their guard is not armed, which today
   they cannot.

Either way the current behaviour costs the predicate's registers and delivers
nothing, so there is no performance argument for the silence.

## Environment

Apple M4 Pro, macOS 26.6.2 (Darwin 25.6.0), Python 3.12.0, triton 3.7.0,
Metal backend.

**Comment by bledden** [comment]
Thanks for flagging this. You're right that the backend cannot use the frontend's debug-off behavior as a reason to discard an assertion that survived into the IR. I removed that assumption. I also found that frameworks retain assertions automatically, so this affects more than handwritten debug kernels.

I took the status-buffer approach you outlined, with an additional requirement: the failing work must stop before the guarded access, not merely report an error after reading or writing out of bounds. Supported retained assertions now produce a host-visible `MetalDeviceAssertionError` before the launch returns, carrying the source message. Masked-out lanes are respected, and a failed launch is not replayed through a fallback. Forms the backend cannot support still refuse, including the retained nested-callee case. A debug-off kernel whose assertion was removed by Triton's frontend does not gain that protection.

The immediate host check has a measured performance cost, which the [release guide](https://github.com/bledden/triton-msl/blob/main/docs/releases/0.3.0.md) discloses. This is backend assertion support, not a claim that I added a hardware trap or ran every complete FlagGems port you mentioned.

Please don't rerun the known-unsafe old out-of-bounds example just to compare versions. If anything still fails, the source, launch configuration and error are enough to start; I can help arrange a bounded witness safely.

**Comment by benkelaya** [comment]
Thanks. Stopping the work before the guarded access, rather than reporting after it, is the right bar. Noted on not rerunning the old out-of-bounds example.
